### ISSUE TITLE
Unit test fails when retrieving vector<map<string, string>>

### ISSUE THREAD

**Issue by markito3** [JeffersonLab/ccdb#74]
- ccdb 1.6.7
- get the following error when running unit tests: [ccdb_unit_test_error.txt](https://github.com/JeffersonLab/ccdb/files/5907833/ccdb_unit_test_error.txt)
- get a similar error with a simple example program: [example_error.txt](https://github.com/JeffersonLab/ccdb/files/5907858/example_error.txt)
with the stack trace shown here: [stack_trace.txt](https://github.com/JeffersonLab/ccdb/files/5907861/stack_trace.txt)
the source code for the simple program is [here](https://github.com/markito3/sweg/blob/master/ccdb/C%2B%2B/example.cc).
- In the simple program, the call to get a vector<map<string, string>> on line 33 is causing the failed assertion. The data it retrieved from the cache in this case, presumably because of the previous fetch of data as a vector of strings.
- If the order of calls is reversed, i.e., the fetch is first performed on the vector of maps and then the vector of strings, the program does not crash.
- I believe something is wrong with the retrieval of data from cache in the original order (first vector of strings, then vector of maps), in particular, the vector of strings containing the column names goes unfilled.

**Comment by markito3** [comment]
Working on v1-main branch. Looking at the unit test test_ccdb_lib. The error reported above still occurs.

The error on the assertion is because an assignment is retrieved from the assignment cache and that assignment does not have the columns loaded. Then that assignment object has its MapData member function called with the columns vector in the signature. Since the columns member vector is zero length (it was never loaded), the assertion fails.

Tried to get around it by invoking scons with "with-cacheon=false". The build succeeded, but now test_ccdb_lib fails as shown in [ccdb_unit_test_nocache.txt](https://github.com/JeffersonLab/ccdb/files/6275220/ccdb_unit_test_nocache.txt), i.e., it cannot find the mysql server. The strange thing is turning off the cache switched the database engine from sqlite to mysql.

**Comment by DraTeots** [comment]
Summary (after the meeting with Mark). 

There are 2 concepts involved. 

- GetAssignment methods has `loadColumns` flag. That is done to lower the number of DB requests when columns information (name and types) are not needed (mos of the times). 
- User API Calibration class has caching, that allows not to go to DB everytime the same data is used. The caching is done via key-value storage. Where key - is the data request. 

It fails because Calibration.cc used cache and doesn't include in cache key the information if columns are loaded or not. So when first the assignment is loaded without columns and then one needs the same assignment with columns CCDB returns it WITHOUT columns from cache. 

The sane fix is to add `loadColumns` information to the cache key

In order to add this to unit tests. The right place for such unit test case is 

```
src/Tests/test_NoMySqlUserAPI.cc
```

and test data 

```
/test/test_vars/test_table:1000:default:
```
Could be used for that. [Here is the data](https://halldweb.jlab.org/cgi-bin/ccdb/show_request?request=/test/test_vars/test_table:1000:default:2017-04-24_10-59-37)

```c++
a = sqliteCalib->GetAssignment("/test/test_vars/test_table:1000:default")
a->GetConstantsTypeTable()->GetColumns()
```

**Comment by markito3** [comment]
Solved with pull request #82 . Also fixes the stand-alone program mentioned in the original post. Closing.
